onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_api_helper.cc

```diff
@@ -0,0 +1,141 @@
+    devices_feature_level = std::max(device.feature_level, devices_feature_level);
+  }
+
+  // nnapi_cpu has the feature 1000
+  if ((devices_feature_level > 0) && (devices_feature_level < target_feature_level)) {
+    LOGS_DEFAULT(INFO) << "Changing NNAPI Feature Level " << target_feature_level
+                       << " to supported by target devices: " << devices_feature_level;
+
+    target_feature_level = static_cast<int32_t>(devices_feature_level);
+  }
+  return target_feature_level;
+}
+
+// get all target devices which satisfy the target_device_option
+// we will always put CPU device at the end if cpu is enabled
+Status GetTargetDevices(const NnApi& nnapi_handle, TargetDeviceOption target_device_option,
+                        InlinedVector<DeviceWrapper>& devices) {
+  // GetTargetDevices is only supported when NNAPI runtime feature level >= ANEURALNETWORKS_FEATURE_LEVEL_3
+  if (GetNNAPIRuntimeFeatureLevel(nnapi_handle) < ANEURALNETWORKS_FEATURE_LEVEL_3)
+    return Status::OK();
+
+  uint32_t num_devices = 0;
+  RETURN_STATUS_ON_ERROR_WITH_NOTE(
+      nnapi_handle.ANeuralNetworks_getDeviceCount(&num_devices), "Getting count of available devices");
+
+  int32_t cpu_index = -1;
+  for (uint32_t i = 0; i < num_devices; i++) {
+    ANeuralNetworksDevice* device = nullptr;
+    const char* device_name = nullptr;
+    int32_t device_type = 0;
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(
+        nnapi_handle.ANeuralNetworks_getDevice(i, &device), "Getting " + std::to_string(i) + "th device");
+
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(nnapi_handle.ANeuralNetworksDevice_getName(device, &device_name),
+                                     "Getting " + std::to_string(i) + "th device's name");
+
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(nnapi_handle.ANeuralNetworksDevice_getType(device, &device_type),
+                                     "Getting " + std::to_string(i) + "th device's type");
+
+    int64_t curr_device_feature_level = 0;
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(nnapi_handle.ANeuralNetworksDevice_getFeatureLevel(device, &curr_device_feature_level),
+                                     "Getting " + std::to_string(i) + "th device's feature level");
+
+    // https://developer.android.com/ndk/reference/group/neural-networks#aneuralnetworksdevice_gettype
+    bool device_is_cpu = device_type == ANEURALNETWORKS_DEVICE_CPU;
+    if ((target_device_option == TargetDeviceOption::CPU_DISABLED && device_is_cpu) ||
+        (target_device_option == TargetDeviceOption::CPU_ONLY && !device_is_cpu)) {
+      continue;
+    }
+
+    if (device_is_cpu) {
+      cpu_index = static_cast<int32_t>(devices.size());
+    }
+    devices.push_back({device, std::string(device_name), device_type, curr_device_feature_level});
+  }
+
+  // put CPU device at the end
+  // 1) it's helpful to accelerate nnapi compile, just assuming nnapi-reference has the lowest priority
+  // and nnapi internally skip the last device if it has already found one.
```

Comment:
just to clarify, is this an assumption on our part, or documenting how the NNAPI implementation works? if the latter, could we also link to their relevant source code?

Reply:
https://source.android.com/docs/core/interaction/neural-networks#cpu-usage

https://android.googlesource.com/platform/frameworks/ml/+/master/nn/runtime/ExecutionPlan.cpp#2303

Reply:
Will put it on the doc.